File: 2023/csharp/day02a/Program.cs
using System;

class Program
{
    const uint TOTAL_RED = 12;
    const uint TOTAL_GREEN = 13;
    const uint TOTAL_BLUE = 14;

    static void Usage()
    {
        string? progPath = Environment.ProcessPath;
        string? progName = Path.GetFileName(progPath);
        Console.WriteLine("usage: " + progName + " <file>");
        System.Environment.Exit(1);
    }

    static uint Process(string contents)
    {
        uint result = 0;

        foreach (string line in contents.Split('\n'))
        {
            if (line.Contains(": "))
            {
                string[] parts = line.Split(": ");
                string gameStr = parts[0];
                string revealsStr = parts[1];

                if (gameStr.Contains(' '))
                {
                    string[] gameParts = gameStr.Split(' ');
                    string gameNumStr = gameParts[1];

                    if (uint.TryParse(gameNumStr, out uint gameNum))
                    {
                        bool valid = true;

                        foreach (string subsetStr in revealsStr.Split("; "))
                        {
                            foreach (string cubesStr in subsetStr.Split(", "))
                            {
                                if (cubesStr.Contains(' '))
                                {
                                    string[] cubeParts = cubesStr.Split(' ');
                                    string amountStr = cubeParts[0];
                                    string color = cubeParts[1];

                                    if (uint.TryParse(amountStr, out uint amount))
                                    {
                                        switch (color)
                                        {
                                            case "red":
                                                if (amount > TOTAL_RED)
                                                {
                                                    valid = false;
                                                }
                                                break;
                                            case "green":
                                                if (amount > TOTAL_GREEN)
                                                {
                                                    valid = false;
                                                }
                                                break;
                                            case "blue":
                                                if (amount > TOTAL_BLUE)
                                                {
                                                    valid = false;
                                                }
                                                break;
                                            default:
                                                throw new Exception("unknown color");
                                        }
                                    }
                                }
                            }
                        }

                        if (valid)
                        {
                            result += gameNum;
                        }
                    }
                }
            }
        }

        return result;
    }

    public static void Main(string[] args)
    {
        if ( args.Length < 1 )
        {
            Program.Usage();
        }
        string fileName = args[0];
        string contents = File.ReadAllText(fileName);
        contents = contents.Remove(contents.Length - 1, 1);
        uint result = Process(contents);
        Console.WriteLine("result = " + result);
    }
}

File: 2023/fsharp/day02a/Program.fs
open System
open System.IO

let TOTAL_RED = 12u
let TOTAL_GREEN = 13u
let TOTAL_BLUE = 14u

let usage () =
    let progname = Path.GetFileName(Environment.GetCommandLineArgs().[0])
    printfn "usage: %s <file>" progname
    Environment.Exit(1)

let processInput (contents: string) =
    let mutable result = 0u
    for line in contents.Split('\n') do
        match line.Split(": ") with
        | [|gameStr; drawsStr|] ->
            match gameStr.Split(' ') with
            | [|_ ; gameNumStr|] ->
                let gameNum = UInt32.Parse(gameNumStr)
                let mutable valid = true
                for drawStr in drawsStr.Split("; ") do
                    for colorAmounts in drawStr.Split(", ") do
                        match colorAmounts.Split(' ') with
                        | [|amountStr; color|] ->
                            let amount = UInt32.Parse(amountStr)
                            if color = "red" && amount > TOTAL_RED then
                                valid <- false
                            elif color = "green" && amount > TOTAL_GREEN then
                                valid <- false
                            elif color = "blue" && amount > TOTAL_BLUE then
                                valid <- false
                        | _ -> ()
                if valid then
                    result <- result + gameNum
            | _ -> ()
        | _ -> ()
    result

[<EntryPoint>]
let main argv =
    if argv.Length < 1 then
        usage()
    let filename = argv.[0]
    let contents = File.ReadAllText(filename)
    let result = processInput contents
    printfn "result = %d" result
    0
